include/deal.II/fe/fe_series.h

```diff
@@ -0,0 +1,428 @@
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__fe_series_H
+#define dealii__fe_series_H
+
+
+
+#include <deal.II/base/std_cxx11/shared_ptr.h>
+#include <deal.II/base/config.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/table.h>
+#include <deal.II/base/table_indices.h>
+#include <deal.II/base/tensor.h>
+#include <deal.II/hp/fe_collection.h>
+#include <deal.II/hp/q_collection.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <vector>
+#include <string>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/*!@addtogroup feall */
+/*@{*/
+
+
+/**
+ * This namespace offers functions to calculate expansion series of the
+ * solution on the reference element. Coefficient of expansion are often used
+ * to estimate local smoothness of the underlying FiniteElement field to decide
+ * on h- or p-adaptive refinement strategy.
+ *
+ * @author Denis Davydov, 2016;
+ */
+namespace FESeries
+{
+  /**
+   * A class to calculate expansion of a scalar FE field into Fourier series
+   * on a reference element. The exponential form of the Fourier series is
+   * based on completeness and Hermitian orthogonality of the set of exponential
+   * functions \f$ \phi_{\bf k}({\bf x}) = \exp(2 \pi i\, {\bf k} \cdot {\bf x})\f$.
+   * For example in 1D the L2-orthogonality condition reads
+   * @f[
+   *   \int_0^1 \phi_k(x) \phi_l^\ast(x) dx=\delta_{kl}.
+   * @f]
+   * Note that \f$ \phi_{\bf k} = \phi_{-\bf k}^\ast \f$.
+   *
+   * The arbitrary scalar FE field on the reference element can be expanded in
+   * the complete orthogonal exponential basis as
+   * @f[
+   *    u({\bf x})
+   *    = \sum_{\bf k} c_{\bf k} \phi_{\bf k}({\bf x}).
+   * @f]
+   * From the orthogonality property of the basis, it follows that
+   * @f[
+   *    c_{\bf k} =
+   *    \int_{[0,1]^d} u({\bf x}) \phi_{\bf k}^\ast ({\bf x}) d{\bf x}\,.
+   * @f]
+   * It is this complex-valued expansion coefficients, that are calculated by
+   * this class. Note that \f$ u({\bf x}) = \sum_i u_i N_i({\bf x})\f$,
+   * where \f$ N_i({\bf x}) \f$ are real-valued FiniteElement shape functions.
+   * Consequently \f$ c_{\bf k} \equiv c_{-\bf k}^\ast \f$ and
+   * we only need to compute \f$ c_{\bf k} \f$ for positive indices
+   * \f$ \bf k \f$ .
+   *
+   * @author Denis Davydov, 2016.
+   */
+  template <int dim>
+  class Fourier : public Subscriptor
```

Comment:
Is there a reason you derive this class (and Legendre further down) from Subscriptor, i.e., is there somewhere where a SmartPointer to this class is stored? (On the other hand, since this class uses a quadrature collection which we sometimes copy, sometimes not, I wonder whether the fe_collection and q_collection arguments should rather be `SmartPointer<...>` to make sure there is no dangling reference.